File: leetcode/archived/Python/99_Recover_Binary_Search_Tree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def recoverTree(self, root) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        cur = root
        pre = first = second = None
        s = []
        while cur:
            if not cur.left:
                if not pre:
                    pre = cur
                if cur.val < pre.val and not first:
                    first = pre
                if first and cur.val > first.val:
                    second = pre
                    break
                pre = cur
                s.append(cur.val)
                cur = cur.right
            else:
                prev = cur.left
                while prev.right and prev.right != cur:
                    prev = prev.right
                if not prev.right:
                    prev.right = cur
                    cur = cur.left
                else:
                    prev.right = None
                    if not pre:
                        pre = cur
                    if cur.val < pre.val and not first:
                        first = pre
                    if first and cur.val > first.val:
                        second = pre
                        break
                    pre = cur
                    s.append(cur.val)
                    cur = cur.right
        if not second:
            second = pre
        first.val, second.val = second.val, first.val


root = TreeNode(68)
root.left = TreeNode(41)
root.left.left = TreeNode(-85)
root.left.left.left = TreeNode(-73)
root.left.left.right = TreeNode(-49)
root.left.left.left.left = TreeNode(-98)
root.left.left.left.left.left = TreeNode(-124)

s = Solution()
res = s.recoverTree(root)

File: leetcode/archived/golang/99_Recover_binary_search_tree.go
package main

type TreeNode struct{
	Val int
	Left *TreeNode
	Right *TreeNode
}

func recoverTree(root *TreeNode) {
	var res []*TreeNode
	inorderTravser(root, &res)
	var pre, n *TreeNode
	pre, n = nil, nil
	for i := 0; i < len(res)-1; i++ {
		if res[i].Val > res[i+1].Val {
			if pre == nil {
				pre, n = res[i], res[i+1]
			} else {
				n = res[i+1]
			}
		}
	}
	pre.Val, n.Val = n.Val, pre.Val

}

func inorderTravser(cur *TreeNode, res *[]*TreeNode) {
	if cur == nil {
		return
	}
	inorderTravser(cur.Left, res)
	*res = append(*res, cur)
	inorderTravser(cur.Right, res)
}

func main() {

}
